app/code/Magento/Cron/Model/ResourceModel/Schedule/ExpressionValidator.php

```diff
@@ -0,0 +1,213 @@
+     */
+    private $fromToModulus;
+
+    /**
+     * @param Schedule $schedule
+     */
+    public function setSchedule($schedule)
+    {
+        $this->schedule = $schedule;
+    }
+
+    /**
+     * @return string
+     */
+    public function getLastErrorMsg()
+    {
+        return $this->lastErrorMsg;
+    }
+
+    /**
+     * Checks the schedule's cron expression validity
+     *
+     * @return bool
+     */
+    public function checkCronExpr()
+    {
+        $expr = $this->schedule->getCronExprArr();
+
+        if (!$expr) {
+            return false;
+        }
+
+        if (sizeof($expr) < self::CRON_EXPRESSION_MIN_PARTS_NUMBER
+            || sizeof($expr) > self::CRON_EXPRESSION_MAX_PARTS_NUMBER
+        ) {
+            return false;
+        }
+
+        $check = $this->checkCronExpressionPart($expr[0])
+            && $this->checkCronExpressionPart($expr[1])
+            && $this->checkCronExpressionPart($expr[2])
+            && $this->checkCronExpressionPart($expr[3])
+            && $this->checkCronExpressionPart($expr[4]);
+
+        return $check;
+    }
+
+    /**
+     * @param string $expr
+     * @return bool
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     */
+    public function checkCronExpressionPart($expr)
+    {
+        $this->lastErrorMsg = '';
+        $this->fromToModulus = [];
+
+        // handle ALL match
+        if ($this->isCronExpressionPartAsterisk($expr)) {
+            return true;
+        }
+
+        // handle multiple options
+        if ($this->isCronExpressionPartList($expr)) {
+            foreach (explode(',', $expr) as $e) {
+                if ($this->checkCronExpressionPart($e)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        // handle modulus
+        $mod = 1;
+        if ($this->isCronExpressionPartModulus($expr)) {
+            $e = explode('/', $expr);
+            if (sizeof($e) !== 2) {
+                $this->lastErrorMsg = __('Invalid cron expression, expecting \'match/modulus\': %1', $expr);
+                return false;
+            }
+
+            if (!is_numeric($e[1])) {
+                $this->lastErrorMsg = __('Invalid cron expression, expecting numeric modulus: %1', $expr);
+                return false;
+            }
+
+            $expr = $e[0];
+            $mod = $e[1];
+        }
+
+        // handle all match by modulus
+        if ($this->isCronExpressionPartAsterisk($expr)) {
+            $fromValue = 0;
+            $toValue = 60;
+        } elseif ($this->isCronExpressionPartRange($expr)) {
+            // handle range
+            $e = explode('-', $expr);
+            if (sizeof($e) !== 2) {
+                $this->lastErrorMsg = __('Invalid cron expression, expecting \'from-to\' structure: %1', $expr);
+                return false;
+            }
+
+            $fromValue = $this->schedule->getNumeric($e[0]);
```

Comment:
`getNumeric` method does not deal with a model state so should not belong to it. To fix this design issue please refactor code:
- extract logic of  `getNumeric` to separate class
- inject it in schedule module and delegate call of `getNumeric` method to created class
- declare `getNumeric` method as `@deprecated`
- inject created class as a dependency into validator
- eliminate schedule model as dependency from validator and pass it as an argument to methods